### PR TITLE
3062: Syndicate campaigns in BPI

### PR DESCRIPTION
**Pull request by rimi-itk**
https://platform.dandigbib.org/issues/3062

This adds handling of campaigns to BPI.

The basic idea is to add an API (hooks) that a module can use to hook into the BPI syndication/push process to add custom content types. This API is then implemented in the `ding_campaign`-module.

### CODE REVIEW

**Comment by GitCop** [comment]
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: d92d6f507f3ab482fd5aa1a026ef2dff9f7d20d2
  * The subject should be kept around 50 characters long. The subject must not be more than 69 characters long. Strive for about 50 characters.

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: d92d6f507f3ab482fd5aa1a026ef2dff9f7d20d2
  * The subject should be kept around 50 characters long. The subject must not be more than 69 characters long. Strive for about 50 characters.

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: d92d6f507f3ab482fd5aa1a026ef2dff9f7d20d2
  * The subject should be kept around 50 characters long. The subject must not be more than 69 characters long. Strive for about 50 characters.

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: d92d6f507f3ab482fd5aa1a026ef2dff9f7d20d2
  * The subject should be kept around 50 characters long. The subject must not be more than 69 characters long. Strive for about 50 characters.

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

File: modules/bpi/bpi.syndicate.inc [hunk 1]
```diff
@@ -682,17 +694,56 @@ function bpi_available_map() {
...
+ * @return string|NULL
+ *   The syndicate url.
+ */
+function bpi_get_syndicate_url($bpi_type, $bpi_id) {
+  $mapping = bpi_get_mapping($bpi_type);
+  if (!empty($mapping)) {
+    $hook = 'bpi_syndicate_action_url';
+    foreach (module_implements($hook) as $module) {
+      $url = module_invoke($module, $hook, $bpi_type, $bpi_id, $mapping);
+      if (!empty($url)) {
```
Comment (kasperg) [inline modules/bpi/bpi.syndicate.inc]: This pattern where we only wants the first hook implementation to run seems to be a common theme. How about implementing a `bpi_module_invoke_one()` function for this.

If it is/becomes a reoccuring pattern throughout the platform then we may move it to a more generic namespace.

File: modules/bpi/translations/da.po [hunk 2]
```diff
@@ -442,3 +442,36 @@ msgstr "Indhold må redigeres"
...
+msgstr "Hvis du ændrer indholdstypen, skal du gemme formularen før du redigerer felttilknytningen."
+
+msgid "Remove mapping of @bpi_type"
+msgstr "Fjern tilknytning af @bpi_type"
+
+msgid "Save to remove mapping of @bpi_type"
+msgstr "Gem for at fjerne tilknytning af @bpi_type"
+
+msgid "Add mapping"
+msgstr "Tilføj tilknytning"
```
Comment (kasperg) [inline modules/bpi/translations/da.po]: :+1: Thanks for updating this.

File: modules/ding_campaign/ding_campaign.bpi.inc [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        '#options' => array('title' => t('Title')),
+        '#value' => 'title',
+        '#disabled' => TRUE,
+      ),
+
+      'bpi_field_body' => array(
+        '#type' => 'select',
+        '#title' => t('BPI body'),
+        '#description' => t('Field to extract the main content from (body field).'),
+        '#options' => $field_options,
```
Comment (kasperg) [inline modules/ding_campaign/ding_campaign.bpi.inc]: When looking at this, `bpi_bpi_get_field_mapping_form()` and the way the form is handled it seems as if a quite specific structure is needed for this to work. This does not go very well with a hook that basicly allows you to provide whatever form API structure you want.

I suggest replacing `bpi_get_field_mapping_form` with a hook which returns an array of BPI field data (names, descriptions + ?) and then that is used to generate the form API.

Example:

```php
function ding_campaign_bpi_get_fields() {
   return array(
     bpi_field_body => array(
       'title' => t('BPI body'),
       'description' => t('Field to extract the main content from (body field).'),
     ),
     // and so on...
   )
}

File: modules/ding_campaign/ding_campaign.bpi.inc [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    );
+  }
+}
+
+/**
+ * Implements hook_bpi_syndicate_form().
+ */
+function ding_campaign_bpi_syndicate_form_alter(&$form, $bpi_content, $syndicated_images) {
+  $current_language = $form['language']['#value'];
+  $data = @json_decode($bpi_content['data'], TRUE);
```
Comment (kasperg) [inline modules/ding_campaign/ding_campaign.bpi.inc]: Why do we want to suppress errors here? That is generally not a good practice.

File: modules/ding_campaign/ding_campaign.bpi.inc [hunk 5]
```diff
@@ -0,0 +1,108 @@
+<?php
+
+/**
+ * @file
+ * Implementation of BPI hooks.
```
Comment (kasperg) [inline modules/ding_campaign/ding_campaign.bpi.inc]: :+1: for a separate file for this.

File: modules/bpi/bpi.api.php [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+/**
+ * Get bpi image type for an image field.
+ *
+ * @param string $image_field_name
+ *   The image field name.
+ * @param object $node
+ *   The node.
+ *
+ * @return string|NULL
+ *   The image type.
```
Comment (kasperg) [inline modules/bpi/bpi.api.php]: I am unsure what image type refers to here - is it a mime type, a Drupal field type or?

File: modules/bpi/bpi.api.php [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+/**
+ * @addtogroup hooks
+ * @{
+ */
+
+/**
+ * Return a list of BPI node types a module can handle.
+ *
+ * @return array
```
Comment (kasperg) [inline modules/bpi/bpi.api.php]: So this is basicly `string[]`.

File: modules/bpi/bpi.api.php [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+function hook_bpi_get_field_mapping_form($bpi_type, $content_type, array $field_mapping) {
+    return array();
+}
+
+/**
+ * Alter the form for defining a BPI node mapping.
+ *
+ * @param string $bpi_type
+ *   The BPI node type.
+ * @param array $form_item
```
Comment (kasperg) [inline modules/bpi/bpi.api.php]: Is there a reason why this parameter is called `$form_item` but referred to as "form"?

File: modules/bpi/bpi.install [hunk 9]
```diff
@@ -210,3 +210,69 @@ function bpi_update_7004(&$sandbox) {
...
+ * Update our own translations.
+ */
+function bpi_update_7005() {
+  bpi_translation_update();
+}
+
+/**
+ * Update the database with the newest translations.
+ */
+function bpi_translation_update() {
```
[the remark this answers is not in the document]
Comment (rimi-itk) [reply modules/bpi/bpi.install]: Bummer!

File: modules/bpi/bpi.admin.inc [hunk 10]
```diff
@@ -130,85 +130,146 @@ function bpi_admin_settings_form_validate(&$form, &$form_state) {
...
+            'wrapper' => 'bpi-wrapper',
+          ),
+          '#required' => TRUE,
+        ),
+
+        'bpi_content_type' => array(
+          '#type' => 'select',
+          '#title' => t('Content type'),
+          '#description' => t('Select a content type into which content from BPI will be syndicated.')
+          .'<br/>' . t('If you change the content type, you have to save the form before editing the field mapping.'),
```
[the remark this answers is not in the document]
Comment (rimi-itk) [reply modules/bpi/bpi.admin.inc]: I'll fix the string.

File: modules/bpi/bpi.api.php (hunk 7 above)
Comment (rimi-itk) [reply modules/bpi/bpi.api.php]: Yes. I'll update the doc to reflect this.

File: modules/bpi/bpi.api.php (hunk 6 above)
Comment (rimi-itk) [reply modules/bpi/bpi.api.php]: I'll clarify in the doc comment.

File: modules/bpi/bpi.api.php (hunk 8 above)
Comment (rimi-itk) [reply modules/bpi/bpi.api.php]: No, it's a typo.

File: modules/ding_campaign/ding_campaign.bpi.inc (hunk 4 above)
Comment (rimi-itk) [reply modules/ding_campaign/ding_campaign.bpi.inc]: You're absolutely right. I don't know why I did that – will fix!

File: modules/bpi/bpi.install [hunk 11]
```diff
@@ -210,3 +210,69 @@ function bpi_update_7004(&$sandbox) {
...
+    $bpi_mappings = array(
+      array(
+        'bpi_node_type' => 'article',
+        'bpi_content_type' => variable_get('bpi_content_type', ''),
+        'bpi_field_mapping' => array(
+          'bpi_field_title' => variable_get('bpi_field_title', ''),
+          'bpi_field_teaser' => variable_get('bpi_field_teaser', ''),
+          'bpi_field_body' => variable_get('bpi_field_body', ''),
+          'bpi_field_tags' => variable_get('bpi_field_tags', ''),
+          'bpi_field_materials' => variable_get('bpi_field_materials', ''),
```
[the remark this answers is not in the document]
Comment (rimi-itk) [reply modules/bpi/bpi.install]: No, there's no reason for this – apart from reusing the old (Drupal) variable names. I'll clean this up.

File: modules/bpi/bpi.bpi.inc [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+function bpi_bpi_syndicate_get_bpi_types() {
+  return array('article');
+}
+
+/**
+ * Implements hook_bpi_get_field_mapping().
+ */
+function bpi_bpi_get_field_mapping_form($bpi_type, $content_type, array $field_mapping) {
+  if (in_array($bpi_type, bpi_bpi_syndicate_get_bpi_types())) {
+    $field_options = array('' => '') + bpi_get_field_instances($content_type);
```
[the remark this answers is not in the document]
Comment (rimi-itk) [reply modules/bpi/bpi.bpi.inc]: Will fix.

File: modules/bpi/bpi.syndicate.inc [hunk 13]
```diff
@@ -682,17 +694,56 @@ function bpi_available_map() {
...
+function bpi_syndicate_action($bpi_type, $bpi_id) {
+  $mapping = bpi_get_mapping($bpi_type);
+
+  if (!empty($mapping) && !empty($mapping['bpi_content_type'])) {
+    $hook = 'bpi_syndicate_action_url';
+    foreach (module_implements($hook) as $module) {
+      $url = module_invoke($module, $hook, $bpi_type, $bpi_id, $mapping);
+      if (!empty($url)) {
+        header('Location: ' . $url, TRUE, 301);
+        drupal_exit($url);
```
[the remark this answers is not in the document]
Comment (rimi-itk) [reply modules/bpi/bpi.syndicate.inc]: Yes, but I don't remember exactly why (maybe something with the query string). However, this function is now obsolete and I'll remove it.

File: modules/bpi/bpi.syndicate.inc [hunk 14]
```diff
@@ -441,10 +447,16 @@ function bpi_preprocess_bpi_search_results(array &$variables) {
...
         array('attributes' => array('class' => 'use-ajax'))
       ),
-      l(t('Syndicate'),
-        'admin/bpi/syndicate/' . $item['bpi_id']
-      ),
     );
+
+    $syndicate_url = bpi_get_syndicate_url($item['type'], $item['bpi_id']);
+    if ($syndicate_url) {
+      $actions[] = '<a href="' . check_plain($syndicate_url) . '">' . t('Syndicate') . '</a>';
```
[the remark this answers is not in the document]
Comment (rimi-itk) [reply modules/bpi/bpi.syndicate.inc]: I couldn't get `l` to stop encoding the query string in the syndicate url, but now I've found out that marking the url as external (by passing `array('external' => TRUE)` in the options) makes it work.

File: modules/bpi/bpi.syndicate.inc (hunk 1 above)
Comment (rimi-itk) [reply modules/bpi/bpi.syndicate.inc]: Good idea!

**Comment by GitCop** [comment]
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 7fd0e2c5475841fcfeb21d27fb19553f9e03f1a3
  * The subject should be kept around 50 characters long. The subject must not be more than 69 characters long. Strive for about 50 characters.

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 7fd0e2c5475841fcfeb21d27fb19553f9e03f1a3
  * The subject should be kept around 50 characters long. The subject must not be more than 69 characters long. Strive for about 50 characters.

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

File: modules/bpi/bpi.admin.inc [hunk 15]
```diff
@@ -130,85 +130,146 @@ function bpi_admin_settings_form_validate(&$form, &$form_state) {
...
-
-  if (isset($form_state['values']['bpi_content_type'])) {
-    $selected_node_type = $form_state['values']['bpi_content_type'];
+  $bpi_types = array();
+  $hook = 'bpi_syndicate_get_bpi_types';
+  foreach (module_implements($hook) as $module) {
+    $types = module_invoke($module, $hook);
+    if (is_array($types)) {
+      $bpi_types = array_merge($bpi_types, $types);
+    }
```
[the remark this answers is not in the document]
Comment (rimi-itk) [reply modules/bpi/bpi.admin.inc]: No, but I had misunderstood how return values are handled by `module_invoke_all`. It does exactly what I need. Thanks for pointing this out!

File: modules/bpi/bpi.module [hunk 16]
```diff
@@ -1014,3 +1027,133 @@ function bpi_xautoload($adapter) {
...
+}
+
+/**
+ * Get mapping of a BPI node type by Drupal content type.
+ *
+ * @param string $content_type
+ *   The Drupal content type.
+ *
+ * @return null|array
+ *   The mapping if it exists. Otherwise null.
```
[the remark this answers is not in the document]
Comment (rimi-itk) [reply modules/bpi/bpi.module]: I'll add documentation on the structure of the mapping.

File: modules/bpi/bpi.module [hunk 17]
```diff
@@ -1014,3 +1027,133 @@ function bpi_xautoload($adapter) {
...
+              'description' => 'Data in this group will only be saved when "Save and push" is used.',
+            ),
+            'formatter' => 'closed',
+          ),
+        );
+        field_group_group_save($group);
+
+        // Enable the field group.
+        ctools_include('export');
+        ctools_export_crud_enable('field_group', $group->identifier);
```
[the remark this answers is not in the document]
Comment (rimi-itk) [reply modules/bpi/bpi.module]: I'll change it (back) to 36 for consistency with the value in `bpi_features.field_group.inc`.

File: modules/bpi/bpi.module [hunk 18]
```diff
@@ -1014,3 +1027,133 @@ function bpi_xautoload($adapter) {
...
+            'module' => 'workflowfield',
+            'settings' => array(
+              'comment' => 1,
+              'name_as_title' => 1,
+            ),
+            'type' => 'workflow_default',
+            'weight' => 35,
+          ),
+        );
+        field_create_instance($instance);
```
[the remark this answers is not in the document]
Comment (rimi-itk) [reply modules/bpi/bpi.module]: Can we find a better way to do this? I'm not an expert in this field.

Will it be a better solution to make the weights depend on current content type fields?

**Comment by DDBdrift** [comment]
https://platform.dandigbib.org/issues/3062 | Closed

**Pull request closed without merging**